### PR TITLE
handle plain response

### LINKED ISSUE #701: NSwag.CodeGeneration.CSharp it is impossible to get response.Content with out json deserialization
Can you support auto/manual/custom response content processing?

It is impossible (ClientTemplate.tt, line: 310) to get raw response content. Generated client code trying to deserialize content as json formatted string and fails.

I have an issue when trying to get response with content type text/html, text/plain.

### LINKED ISSUE #731: Handle simple non-json calls
Not sure where the culprit is, but the definition below ended up generating code that assumes data is fully in json format, and it failed when trying to retrieve simple 

In the generated code, I had to do set the media type to text/plain for both the request and response, and also remove the calls to serialize to / deserialize from json.

This comes from the TeamCity REST API swagger definition.

```
        "put":{
            "tags":[
               "Project"
            ],
            "operationId":"setProjectFiled",
            "parameters":[
               {
                  "name":"projectLocator",
                  "in":"path",
                  "required":true,
                  "type":"string"
               },
               {
                  "name":"field",
                  "in":"path",
                  "required":true,
                  "type":"string"
               },
               {
                  "in":"body",
                  "name":"body",
                  "required":false,
                  "schema":{
                     "type":"string"
                  }
               }
            ],
            "responses":{
               "200":{
                  "description":"successful operation",
                  "schema":{
                     "type":"string"
                  },
                  "headers":{

                  }
               }
            }
         }
      },
```

Here is the code with fixes
```
 public async System.Threading.Tasks.Task<string> SetProjectFiledAsync(string projectLocator, string field,
            string body, System.Threading.CancellationToken cancellationToken)
        {
            if (projectLocator == null)
                throw new System.ArgumentNullException("projectLocator");

            if (field == null)
                throw new System.ArgumentNullException("field");

            var urlBuilder_ = new System.Text.StringBuilder();
            urlBuilder_.Append(BaseUrl).Append("/app/rest/projects/{projectLocator}/{field}");
            urlBuilder_.Replace("{projectLocator}", System.Uri.EscapeDataString(projectLocator.ToString()));
            urlBuilder_.Replace("{field}", System.Uri.EscapeDataString(field.ToString()));

var client_ = Client_;
            try
            {
                using (var request_ = new System.Net.Http.HttpRequestMessage())
                {
                    PrepareRequest(client_, urlBuilder_);
                    var url_ = urlBuilder_.ToString();
                    PrepareRequest(client_, url_);

                    var content_ = new System.Net.Http.StringContent(body);
                    content_.Headers.ContentType.MediaType = "text/plain";
                    request_.Content = content_;
                    request_.Method = new System.Net.Http.HttpMethod("PUT");
                    request_.RequestUri = new System.Uri(url_, System.UriKind.RelativeOrAbsolute);
                    request_.Headers.Accept.Add(
                        new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("text/plain"));
                    var response_ =
                        await client_.SendAsync(request_, System.Net.Http.HttpCompletionOption.ResponseHeadersRead,
                            cancellationToken).ConfigureAwait(false);
                    try
                    {
                        var headers_ = System.Linq.Enumerable.ToDictionary(response_.Headers, h_ => h_.Key,
                            h_ => h_.Value);
                        foreach (var item_ in response_.Content.Headers)
                            headers_[item_.Key] = item_.Value;

                        ProcessResponse(client_, response_);

                        var status_ = ((int) response_.StatusCode).ToString();
                        if (status_ == "200")
                        {
                            var responseData_ = await response_.Content.ReadAsStringAsync().ConfigureAwait(false);
                            return responseData_;
                        }
                        else if (status_ != "200" && status_ != "204")
                        {
                            var responseData_ = await response_.Content.ReadAsStringAsync().ConfigureAwait(false);
                            throw new SwaggerException(
                                "The HTTP status code of the response was not expected (" + (int) response_.StatusCode +
                                ").", status_, responseData_, headers_, null);
                        }

                        return default(string);
                    }
                    finally
                    {
                        if (response_ != null)
                            response_.Dispose();
                    }
                }
            }
            finally
            {
                if (client_ != null)
                    client_.Dispose();
            }
        }
```

Here is the original incorrect code
```
        public async System.Threading.Tasks.Task<string> SetProjectFiledAsync(string projectLocator, string field,
            string body, System.Threading.CancellationToken cancellationToken)
        {
            if (projectLocator == null)
                throw new System.ArgumentNullException("projectLocator");

            if (field == null)
                throw new System.ArgumentNullException("field");

            var urlBuilder_ = new System.Text.StringBuilder();
            urlBuilder_.Append(BaseUrl).Append("/app/rest/projects/{projectLocator}/{field}");
            urlBuilder_.Replace("{projectLocator}", System.Uri.EscapeDataString(projectLocator.ToString()));
            urlBuilder_.Replace("{field}", System.Uri.EscapeDataString(field.ToString()));

var client_ = Client_;
            try
            {
                using (var request_ = new System.Net.Http.HttpRequestMessage())
                {
                    PrepareRequest(client_, urlBuilder_);
                    var url_ = urlBuilder_.ToString();
                    PrepareRequest(client_, url_);

                    var content_ = new System.Net.Http.StringContent(Newtonsoft.Json.JsonConvert.SerializeObject(body));
                    content_.Headers.ContentType.MediaType = "application/json";
                    request_.Content = content_;
                    request_.Method = new System.Net.Http.HttpMethod("PUT");
                    request_.RequestUri = new System.Uri(url_, System.UriKind.RelativeOrAbsolute);
                    request_.Headers.Accept.Add(
                        new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/json"));
                    var response_ =
                        await client_.SendAsync(request_, System.Net.Http.HttpCompletionOption.ResponseHeadersRead,
                            cancellationToken).ConfigureAwait(false);
                    try
                    {
                        var headers_ = System.Linq.Enumerable.ToDictionary(response_.Headers, h_ => h_.Key,
                            h_ => h_.Value);
                        foreach (var item_ in response_.Content.Headers)
                            headers_[item_.Key] = item_.Value;

                        ProcessResponse(client_, response_);

                        var status_ = ((int) response_.StatusCode).ToString();
                        if (status_ == "200")
                        {
                            var responseData_ = await response_.Content.ReadAsStringAsync().ConfigureAwait(false);
                            var result_ = default(string);
                            try
                            {
                                result_ = Newtonsoft.Json.JsonConvert.DeserializeObject<string>(responseData_);
                                return result_;
                            }
                            catch (System.Exception exception)
                            {
                                throw new SwaggerException("Could not deserialize the response body.", status_,
                                    responseData_, headers_, exception);
                            }
                        }
                        else if (status_ != "200" && status_ != "204")
                        {
                            var responseData_ = await response_.Content.ReadAsStringAsync().ConfigureAwait(false);
                            throw new SwaggerException(
                                "The HTTP status code of the response was not expected (" + (int) response_.StatusCode +
                                ").", status_, responseData_, headers_, null);
                        }

                        return default(string);
                    }
                    finally
                    {
                        if (response_ != null)
                            response_.Dispose();
                    }
                }
            }
            finally
            {
                if (client_ != null)
                    client_.Dispose();
            }
        }
```

Not that I also had to change the client definition but that's just for integrated and basic auth, and not related to this issue.

After the code fix, everything works fine.

### PR DESCRIPTION
**Pull request by et1975**
Fixes #731 and fixes #701

### CODE REVIEW

File: src/NSwag.CodeGeneration.CSharp/Templates/Client.Class.ProcessResponse.liquid [hunk 1]
```diff
@@ -14,7 +14,11 @@ var responseData_ = response_.Content == null ? null : await response_.Content.R
 var result_ = default({{ response.Type }}); 
 try
 {
-    result_ = Newtonsoft.Json.JsonConvert.DeserializeObject<{{ response.Type }}>(responseData_, _settings.Value);
+{%    if response.IsPlain -%}
+	result_ = ({{ response.Type }})System.Convert.ChangeType(responseData_, typeof({{ response.Type }}));
```
Comment (RSuter) [inline src/NSwag.CodeGeneration.CSharp/Templates/Client.Class.ProcessResponse.liquid]: Start with 4 spaces.

File: src/NSwag.CodeGeneration/Models/ResponseModelBase.cs [hunk 2]
```diff
@@ -71,6 +71,9 @@ public bool IsDate
             }
         }
 
+        /// <summary>Gets a value indicating whether this is a text/plain response.</summary>
+        public bool IsPlain => _response.Content.ContainsKey("text/plain") || _operationModel.Produces == "text/plain";
```
Comment (RSuter) [inline src/NSwag.CodeGeneration/Models/ResponseModelBase.cs]: Maybe rename to IsPlainText?

File: src/NSwag.CodeGeneration/Models/IOperationModel.cs [hunk 3]
```diff
@@ -15,5 +15,8 @@ public interface IOperationModel
     {
         /// <summary>Gets the responses.</summary>
         IEnumerable<ResponseModelBase> Responses { get; }
+
+        /// <summary>Gets Swagger operation's mime type.</summary>
+        string Produces { get; }
```
Comment (RSuter) [inline src/NSwag.CodeGeneration/Models/IOperationModel.cs]: Where is this implemented?

File: src/NSwag.CodeGeneration/Models/ResponseModelBase.cs (hunk 2 above)
Comment (RSuter) [reply src/NSwag.CodeGeneration/Models/ResponseModelBase.cs]: Does this add support for Swagger 2.0 and OAI 3.0?

File: src/NSwag.CodeGeneration/Models/IOperationModel.cs (hunk 3 above)
Comment (et1975) [reply src/NSwag.CodeGeneration/Models/IOperationModel.cs]: It's already implemented by every class that implements this interface.

**Comment by RSuter** [comment]
Thanks for the PR. I added some comments.

File: src/NSwag.CodeGeneration/Models/ResponseModelBase.cs (hunk 2 above)
Comment (et1975) [reply src/NSwag.CodeGeneration/Models/ResponseModelBase.cs]: the tests pass for both and I'm using 2 ways of specifying the response content type: inline (in 200 response) and top-level "produces". Does it cover all the angles? - I'm not sure, but it's a start.

**Comment by et1975** [comment]
Pushed the requested changes.

**Comment by RSuter** [comment]
Looks good. Thanks for the pr. Release tomorrow

**Comment by RSuter** [comment]
v12.0.4

**Comment by et1975** [comment]
Thank you!

**Pull request merged**